tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,941 @@
+public:
+  SKDObject(SKDObject const&) = delete;
+  SKDObject &operator=(SKDObject const&) = delete;
+
+  ObjectKind getKind() const { return Kind; }
+
+  virtual sourcekitd_variant_type_t getVariantType() = 0;
+
+  virtual void set(sourcekitd_uid_t Key, SKDObjectRef Value) {}
+  virtual void set(size_t Index, SKDObjectRef Value) {}
+
+  virtual SKDObjectRef get(sourcekitd_uid_t Key) { return nullptr; }
+
+  virtual SKDObjectRef get(size_t Index) { return nullptr; }
+  virtual size_t getCount() { return 0; }
+
+  virtual sourcekitd_uid_t getUID() { return nullptr; }
+  virtual Optional<int64_t> getInt64() { return Optional<int64_t>(); }
+  virtual Optional<StringRef> getString() { return Optional<StringRef>(); }
+  virtual bool getBool() { return false; }
+};
+
+class SKDDictionary: public SKDObject {
+public:
+  SKDDictionary() : SKDObject(OK_Dictionary) {}
+  SKDDictionary(SKDDictionary const&) = delete;
+  SKDDictionary &operator=(SKDDictionary const&) = delete;
+
+  sourcekitd_variant_type_t getVariantType() override {
+    return SOURCEKITD_VARIANT_TYPE_DICTIONARY;
+  }
+
+  void set(sourcekitd_uid_t Key, SKDObjectRef Value) override {
+    Storage[Key] = Value;
+  }
+
+  SKDObjectRef get(sourcekitd_uid_t Key) override {
+    auto it = Storage.find(Key);
+    return it != Storage.end() ? it->second : nullptr;
+  }
+
+  bool apply(llvm::function_ref<bool(sourcekitd_uid_t, SKDObjectRef)> Applier) {
+    for (const auto& kv : Storage) {
+      if (!Applier(kv.first, kv.second))
+        return false;
+    }
+    return true;
+  }
+
+  static bool classof(const SKDObject *O) {
+    return O->getKind() == OK_Dictionary;
+  }
+private:
+  std::map<sourcekitd_uid_t, SKDObjectRef> Storage;
+};
+
+class SKDArray: public SKDObject {
+public:
+  SKDArray() : SKDObject(OK_Array) {}
+  SKDArray(SKDArray const&) = delete;
+  SKDArray &operator=(SKDArray const&) = delete;
+
+  sourcekitd_variant_type_t getVariantType() override {
+    return SOURCEKITD_VARIANT_TYPE_ARRAY;
+  }
+
+  void set(size_t Index, SKDObjectRef Value) override {
+    if (Index == SOURCEKITD_ARRAY_APPEND) {
+      Storage.push_back(Value);
+    } else {
+      Storage[Index] = Value;
+    }
+  }
+
+  SKDObjectRef get(size_t Index) override {
+    return Storage[Index];
+  }
+
+  size_t getCount() override {
+    return Storage.size();
+  }
+
+  bool apply(llvm::function_ref<bool(size_t, SKDObjectRef)> Applier) {
+    for (size_t i = 0; i < Storage.size(); ++i) {
+      if (!Applier(i, Storage[i]))
+        return false;
+    }
+    return true;
+  }
+
+  static bool classof(const SKDObject *O) {
+    return O->getKind() == OK_Array;
+  }
+private:
+  std::vector<SKDObjectRef> Storage;
+};
+
+class SKDString: public SKDObject {
+public:
+  SKDString(std::string &&Value) : SKDObject(OK_String), Storage(Value) {}
+
+  sourcekitd_variant_type_t getVariantType() override {
+    return SOURCEKITD_VARIANT_TYPE_STRING;
+  }
+
+  Optional<StringRef> getString() override {
```

Comment:
I suggest also adding `Optional<const char *> getCString()`, it will be useful later on.